File: framework/dotnet/BetterPortal/Handler.cs
using AnyVali;
using System.Collections.Frozen;

namespace BetterPortal;

public sealed class HandlerInputException(string field, ValidationError cause) : Exception("Invalid request " + field, cause)
{
    public int Status => 400;
    public string Field { get; } = field;
    public ValidationError Validation { get; } = cause;
}
public sealed class HandlerOutputException(ValidationError cause) : Exception("Response validation failed", cause)
{
    public int Status => 500;
}

public sealed class ResponseState
{
    private int status = 200;
    private readonly List<KeyValuePair<string, string>> headers = [];
    public int Status { get => status; set { _ = new RawResponse(status: value); status = value; } }
    public IReadOnlyList<KeyValuePair<string, string>> Headers => headers.AsReadOnly();
    public void SetHeader(string name, string value, bool append = false)
    {
        _ = new RawResponse(headers: [new(name, value)]);
        if (name.Equals("content-encoding", StringComparison.OrdinalIgnoreCase)) throw new ArgumentException("Content-Encoding requires a raw response");
        if (!append) RemoveHeader(name);
        headers.Add(new(name, value));
    }
    public void RemoveHeader(string name) => headers.RemoveAll(pair => pair.Key.Equals(name, StringComparison.OrdinalIgnoreCase));
}

/// <summary>Hosting supplies an already resolved scope and verified caller.</summary>
public sealed record RequestContext(ScopedContext Scope, AuthorizedCaller Caller, string Method, string Path,
    IReadOnlyDictionary<string, object?>? Config = null, Generated.MultipartRequest? Multipart = null)
{
    public ResponseState Response { get; init; } = new();
    public Urls Urls { get; init; } = new(Scope, null, null, Path);
    internal RequestClients? ClientContext { get; init; }
    [System.Text.Json.Serialization.JsonIgnore]
    public CancellationToken SnapshotRetired { get; init; }
    [System.Text.Json.Serialization.JsonIgnore]
    public RequestClients Clients => ClientContext ?? throw new InvalidOperationException("Clients require a service request context");
}

public sealed record HandlerContext<TParams, TQuery, THeaders, TBody>(RequestContext RequestContext,
    TParams Params, TQuery Query, THeaders Headers, TBody Request, CancellationToken Cancellation)
{
    public ResponseState Response => RequestContext.Response;
    public Urls Urls => RequestContext.Urls;
}
public sealed record Invocation(object? Value, object? Params, object? Query);

/// <summary>Common registration surface for handlers with different input/output types.</summary>
public abstract class Handler
{
    public abstract Schema? ResponseSchema { get; }
    public virtual bool IsRaw => false;
    public virtual bool IsStreaming => false;
    public virtual IReadOnlyCollection<string> StreamRendererKeys => [];
    public virtual Dictionary<string, object?>? StreamingMetadata => null;
    public virtual RawResponse OpenStream(RequestContext context, IReadOnlyDictionary<string, object?> values, bool sse = false,
        Func<string, object?, object?, RenderContext>? renderContext = null, CancellationToken cancellation = default)
        => throw new InvalidOperationException("Handler does not stream");
    public virtual ValueTask<string?> StreamShell(RequestContext context, IReadOnlyDictionary<string, object?> values, string connectionPath, string mode,
        Func<string, object?, object?, RenderContext> renderContext, CancellationToken cancellation = default) => ValueTask.FromResult<string?>(null);
    public virtual IReadOnlyList<Renderer> Renderers => [];
    public abstract IReadOnlyDictionary<string, Schema> Schemas { get; }
    public Dictionary<string, object?> InputDocument => Contracts.ObjectDocument(new[] { "params", "query", "headers", "request" }.ToDictionary(name => name,
        name => (Dictionary<string, object?>)Json.Read(Json.Write(V.Export(Schemas.GetValueOrDefault(name, Contracts.Get("JsonObjectSchema")))))!), "reject");
    internal abstract ValueTask<Invocation> ExecuteBoxed(RequestContext context, IReadOnlyDictionary<string, object?> values, CancellationToken cancellation);
    internal async ValueTask<object?> InvokeBoxed(RequestContext context, IReadOnlyDictionary<string, object?> values, CancellationToken cancellation) =>
        (await ExecuteBoxed(context, values, cancellation)).Value;
}

public abstract class Handler<TParams, TQuery, THeaders, TBody> : Handler
{
    public override IReadOnlyDictionary<string, Schema> Schemas { get; }
    protected Handler(Schema? @params, Schema? query, Schema? headers, Schema? request)
    {
        Schemas = new Dictionary<string, Schema?>(StringComparer.Ordinal)
        { ["params"] = @params, ["query"] = query, ["headers"] = headers, ["request"] = request }
            .Where(pair => pair.Value is not null).ToFrozenDictionary(pair => pair.Key, pair => pair.Value!, StringComparer.Ordinal);
    }
    private T Parse<T>(string name, IReadOnlyDictionary<string, object?> values)
    {
        try { return Contracts.Parse<T>(Schemas.GetValueOrDefault(name, Contracts.Get("JsonObjectSchema")), values.GetValueOrDefault(name, new Dictionary<string, object?>())); }
        catch (ValidationError error) { throw new HandlerInputException(name, error); }
    }
    public HandlerContext<TParams, TQuery, THeaders, TBody> Prepare(RequestContext context, IReadOnlyDictionary<string, object?> values, CancellationToken cancellation = default) =>
        new(context, Parse<TParams>("params", values), Parse<TQuery>("query", values), Parse<THeaders>("headers", values), Parse<TBody>("request", values), cancellation);
}

public sealed class Handler<TParams, TQuery, THeaders, TBody, TResult> : Handler<TParams, TQuery, THeaders, TBody>
{
    public override Schema ResponseSchema { get; }
    public override IReadOnlyList<Renderer> Renderers { get; }
    private readonly Func<HandlerContext<TParams, TQuery, THeaders, TBody>, ValueTask<TResult>> run;
    public Handler(Schema response, Func<HandlerContext<TParams, TQuery, THeaders, TBody>, ValueTask<TResult>> run,
        Schema? @params = null, Schema? query = null, Schema? headers = null, Schema? request = null,
        IEnumerable<Renderer<TResult>>? renderers = null) : base(@params, query, headers, request)
    {
        ArgumentNullException.ThrowIfNull(response); ArgumentNullException.ThrowIfNull(run);
        ResponseSchema = response; this.run = run;
        Renderers = Renderer.Unique(renderers ?? []);
    }

    private async ValueTask<(TResult Value, HandlerContext<TParams, TQuery, THeaders, TBody> Prepared)> Execute(RequestContext context, IReadOnlyDictionary<string, object?> values, CancellationToken cancellation)
    {
        cancellation.ThrowIfCancellationRequested();
        var prepared = Prepare(context, values, cancellation);
        var result = await run(prepared).AsTask().WaitAsync(cancellation);
        if (result is RawResponse raw)
        {
            await raw.DisposeAsync();
            throw new InvalidOperationException("Raw responses require a RawHandler");
        }
        cancellation.ThrowIfCancellationRequested();
        try { return (Contracts.Parse<TResult>(ResponseSchema, result), prepared); }
        catch (ValidationError error) { throw new HandlerOutputException(error); }
    }
    public async ValueTask<TResult> Invoke(RequestContext context, IReadOnlyDictionary<string, object?> values, CancellationToken cancellation = default) =>
        (await Execute(context, values, cancellation)).Value;
    internal override async ValueTask<Invocation> ExecuteBoxed(RequestContext context, IReadOnlyDictionary<string, object?> values, CancellationToken cancellation)
    {
        var result = await Execute(context, values, cancellation);
        return new(result.Value, result.Prepared.Params, result.Prepared.Query);
    }
}

File: framework/python/betterportal/handler.py
"""Typed handlers; hosting supplies an already resolved and authorized context."""
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
import inspect
from types import MappingProxyType
from typing import Any, Awaitable, Callable, Generic, Iterable, Mapping, TYPE_CHECKING, TypeVar, cast

import anyvali as av
from .authorization import AuthorizedCaller
from .context import ScopedContext
from .contracts import contract, export, object_document
from .generated_types import HttpMethod, MultipartRequest
from .urls import Urls
from .bp_headers import BpHeaders
from .observability import TraceContext, current_trace, current_observability
if TYPE_CHECKING:
    from .rendering import Renderer
    from .clients import RequestClients

Params = TypeVar("Params")
Query = TypeVar("Query")
Headers = TypeVar("Headers")
Body = TypeVar("Body")
Result = TypeVar("Result")


class HandlerInputError(ValueError):
    status = 400
    def __init__(self, field: str, cause: av.ValidationError):
        super().__init__("Invalid request " + field)
        self.field, self.validation = field, cause


class HandlerOutputError(ValueError):
    status = 500


class ResponseState:
    """Per-request status and application headers; transport/CORS stay host-owned."""
    def __init__(self) -> None: self._status = 200; self._headers: list[tuple[str, str]] = []; self.bp_headers = BpHeaders()
    @property
    def status(self) -> int: return self._status
    @status.setter
    def status(self, value: int) -> None:
        from .response import RawResponse
        RawResponse(status=value)
        self._status = value
    @property
    def headers(self) -> tuple[tuple[str, str], ...]: return (*self._headers, *self.bp_headers.emit())
    def set_header(self, name: str, value: str, *, append: bool = False) -> None:
        from .response import RawResponse
        RawResponse(headers=[(name, value)])
        if name.lower() in ("content-encoding", "cache-control"): raise ValueError("Response header requires a raw response")
        if not append: self.remove_header(name)
        self._headers.append((name, value))
    def remove_header(self, name: str) -> None:
        self._headers = [(key, value) for key, value in self._headers if key.lower() != name.lower()]


@dataclass(frozen=True)
class RequestContext:
    scope: ScopedContext
    caller: AuthorizedCaller
    method: HttpMethod
    path: str
    config: Mapping[str, Any] = field(default_factory=dict)
    multipart: MultipartRequest | None = None
    response: ResponseState = field(default_factory=ResponseState)
    url_context: Urls | None = None
    client_context: RequestClients | None = field(default=None, repr=False, compare=False)
    _retired: asyncio.Event | None = field(default=None, repr=False, compare=False)
    @property
    def trace(self) -> TraceContext | None: return current_trace()

    def require_elevation(self, requirement: Mapping[str, Any]) -> None:
        from .elevation import require_elevation, ElevationRequired
        from .security import TokenError
        from .service import RequestError
        if self.caller.mode == "service": raise RequestError(403, "Human authentication required", scope=self.scope)
        try: require_elevation(self.caller.user, requirement)
        except ElevationRequired as error: raise RequestError(401, str(error), error.headers, scope=self.scope) from error
        except TokenError as error: raise RequestError(401, str(error), scope=self.scope) from error
    @property
    def urls(self) -> Urls: return self.url_context or Urls(self.scope, None, None, self.path)
    @property
    def clients(self) -> RequestClients:
        if self.client_context is None: raise RuntimeError("Clients require a service request context")
        return self.client_context


@dataclass(frozen=True)
class HandlerContext(Generic[Params, Query, Headers, Body]):
    request_context: RequestContext
    params: Params
    query: Query
    headers: Headers
    request: Body
    def require_elevation(self, requirement: Mapping[str, Any]) -> None: self.request_context.require_elevation(requirement)
    async def webhook(self, event_id: str, payload: Any, *, idempotency_key: str | None = None) -> str:
        return await self.request_context.clients.webhook(event_id, payload, idempotency_key=idempotency_key)
    @property
    def tenant(self): return self.request_context.scope.tenant
    @property
    def app(self): return self.request_context.scope.app
    @property
    def user(self): return self.request_context.caller.user
    @property
    def service_caller(self): return self.request_context.caller.service
    @property
    def caller_mode(self): return self.request_context.caller.mode
    @property
    def config(self): return self.request_context.config
    @property
    def method(self): return self.request_context.method
    @property
    def path(self): return self.request_context.path
    @property
    def multipart(self): return self.request_context.multipart
    @property
    def clients(self): return self.request_context.clients
    @property
    def bp_headers(self): return self.response.bp_headers
    @property
    def trace(self): return self.request_context.trace
    @property
    def obs(self): return current_observability()
    def require_permission(self, service_id: str, view_id: str, action: str) -> str:
        from .auth_helpers import require_permission
        return require_permission(self.request_context, service_id, view_id, action)
    @property
    def response(self) -> ResponseState: return self.request_context.response
    @property
    def urls(self) -> Urls: return self.request_context.urls


@dataclass(frozen=True)
class Invocation(Generic[Result]):
    value: Result
    params: Any
    query: Any


class HandlerInputs(Generic[Params, Query, Headers, Body]):
    def __init__(self, *, params: av.BaseSchema[Params] | None = None, query: av.BaseSchema[Query] | None = None,
                 headers: av.BaseSchema[Headers] | None = None, request: av.BaseSchema[Body] | None = None):
        if any(schema is not None and not isinstance(schema, av.BaseSchema) for schema in (params, query, headers, request)):
            raise TypeError("Handler inputs must use AnyVali schemas")
        self.schemas: Mapping[str, av.BaseSchema[Any]] = MappingProxyType({name: schema for name, schema in
            (("params", params), ("query", query), ("headers", headers), ("request", request)) if schema is not None})

    @property
    def input_document(self) -> dict[str, Any]:
        # The same documents feed native type generation and runtime validation.
        return object_document({name: export(self.schemas.get(name, contract("JsonObjectSchema")))
            for name in ("params", "query", "headers", "request")}, unknown_keys="reject")

    def prepare(self, context: RequestContext, values: Mapping[str, Any]) -> HandlerContext[Params, Query, Headers, Body]:
        parsed = {}
        for name in ("params", "query", "headers", "request"):
            try:
                parsed[name] = self.schemas.get(name, contract("JsonObjectSchema")).parse(values.get(name, {}))
            except av.ValidationError as error:
                raise HandlerInputError(name, error) from error
        return HandlerContext(context, parsed["params"], parsed["query"], parsed["headers"], parsed["request"])

class Handler(HandlerInputs[Params, Query, Headers, Body], Generic[Params, Query, Headers, Body, Result]):
    is_raw = False

    def __init__(self, response: av.BaseSchema[Result],
                 run: Callable[[HandlerContext[Params, Query, Headers, Body]], Result | Awaitable[Result]], *,
                 params: av.BaseSchema[Params] | None = None, query: av.BaseSchema[Query] | None = None,
                 headers: av.BaseSchema[Headers] | None = None, request: av.BaseSchema[Body] | None = None,
                 renderers: Iterable[Renderer[Result]] = ()):
        super().__init__(params=params, query=query, headers=headers, request=request)
        if not isinstance(response, av.BaseSchema) or not callable(run):
            raise TypeError("An AnyVali response schema and handler function are required")
        self.response_schema, self.run = response, run
        from .rendering import renderers as unique_renderers
        self.renderers = unique_renderers(renderers)

    async def execute(self, context: RequestContext, values: Mapping[str, Any]) -> Invocation[Result]:
        prepared = self.prepare(context, values)
        result = self.run(prepared)
        if inspect.isawaitable(result):
            result = await result
        from .response import RawResponse
        if isinstance(result, RawResponse):
            await result.aclose()
            raise TypeError("Raw responses require a RawHandler")
        try:
            return Invocation(self.response_schema.parse(cast(Result, result)), prepared.params, prepared.query)
        except av.ValidationError as error:
            raise HandlerOutputError("Response validation failed") from error

    async def invoke(self, context: RequestContext, values: Mapping[str, Any]) -> Result:
        return (await self.execute(context, values)).value
